id/id.1
.Dd December 2008
.Dt ID 1

.Sh NAME

.Nm id
.Nd return user identity

.Sh SYNOPSIS

.Nm id
.Ar [user]

.Nm id
.Ar -G[-n] [user]

.Nm id
.Ar -g[-nr] [user]

.Nm id
.Ar -u[-nr] [user]

.Sh DESCRIPTION

    If no user operand is provided, the id utility shall write the user and
group IDs and the corresponding user and group names of the invoking process
to standard output. If the effective and real IDs do not match, both shall be
written. If multiple groups are supported by the underlying system (see the
description of {NGROUPS_MAX} in the System Interfaces volume of IEEE Std
1003.1-2001), the supplementary group affiliations of the invoking process
shall also be written.

    If a user operand is provided and the process has the appropriate
privileges, the user and group IDs of the selected user shall be written. In
this case, effective IDs shall be assumed to be identical to real IDs. If the
selected user has more than one allowable group membership listed in the
group database, these shall be written in the same manner as the
supplementary groups described in the preceding paragraph.

.Sh OPTIONS

    The id utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -G
        Output all different group IDs (effective, real, and supplementary)
only, using the format "%u\n". If there is more than one distinct group
affiliation, output each such affiliation, using the format " %u", before the
<newline> is output.
    -g
        Output only the effective group ID, using the format "%u\n".
    -n
        Output the name in the format "%s" instead of the numeric ID using
the format "%u".
    -r
        Output the real ID instead of the effective ID.
    -u
        Output only the effective user ID, using the format "%u\n".

.Sh OPERANDS

    The following operand shall be supported:

    user
        The login name for which information is to be written.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of id:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error and informative
messages written to standard output.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    The following formats shall be used when the LC_MESSAGES locale category
specifies the POSIX locale. In other locales, the strings uid, gid, euid,
egid, and groups may be replaced with more appropriate strings corresponding
to the locale.

    "uid=%u(%s) gid=%u(%s)\n", <real user ID>, <user-name>,
        <real group ID>, <group-name>

    If the effective and real user IDs do not match, the following shall be
inserted immediately before the '\n' character in the previous format:

    " euid=%u(%s)"

    with the following arguments added at the end of the argument list:

    <effective user ID>, <effective user-name>

    If the effective and real group IDs do not match, the following shall be
inserted directly before the '\n' character in the format string (and after
any addition resulting from the effective and real user IDs not matching):

    " egid=%u(%s)"

    with the following arguments added at the end of the argument list:

    <effective group-ID>, <effective group name>

    If the process has supplementary group affiliations or the selected user
is allowed to belong to multiple groups, the first shall be added directly
before the <newline> in the format string:

    " groups=%u(%s)"

    with the following arguments added at the end of the argument list:

    <supplementary group ID>, <supplementary group name>

    and the necessary number of the following added after that for any
remaining supplementary group IDs:

    ",%u(%s)"

    and the necessary number of the following arguments added at the end of
the argument list:

    <supplementary group ID>, <supplementary group name>

    If any of the user ID, group ID, effective user ID, effective group ID,
or supplementary/multiple group IDs cannot be mapped by the system into
printable user or group names, the corresponding "(%s)" and name argument
shall be omitted from the corresponding format string.

    When any of the options are specified, the output format shall be as
described in the OPTIONS section.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    Output produced by the -G option and by the default case could
potentially produce very long lines on systems that support large numbers of
supplementary groups. (On systems with user and group IDs that are 32-bit
integers and with group names with a maximum of 8 bytes per name, 93
supplementary groups plus distinct effective and real group and user IDs
could theoretically overflow the 2048-byte {LINE_MAX} text file line limit on
the default output case. It would take about 186 supplementary groups to
overflow the 2048-byte barrier using id -G). This is not expected to be a
problem in practice, but in cases where it is a concern, applications should
consider using fold -s before postprocessing the output of id.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    The functionality provided by the 4 BSD groups utility can be simulated
using:

    id -Gn [ user ]

    The 4 BSD command groups was considered, but it was not included because
it did not provide the functionality of the id utility of the SVID. Also, it
was thought that it would be easier to modify id to provide the additional
functionality necessary to systems with multiple groups than to invent
another command.

    The options -u, -g, -n, and -r were added to ease the use of id with
shell commands substitution. Without these options it is necessary to use
some preprocessor such as sed to select the desired piece of information.
Since output such as that produced by:

    id -u -n

    is frequently wanted, it seemed desirable to add the options.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    fold, logname, who, the System Interfaces volume of IEEE Std 1003.1-2001,
getgid(), getgroups(), getuid()



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



id.c

/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <argp.h>
#include <libpu.h>

static const char doc[] =
N_("id - return user identity");

static const char args_doc[] = N_("[user]");
static struct argp_option options[] = {
	{ NULL, 'g', NULL, 0,
	  N_("Output only the effective group ID") },
	{ NULL, 'G', NULL, 0,
	  N_("Output all different group IDs") },
	{ NULL, 'n', NULL, 0,
	  N_("Output the name instead of the numeric ID") },
	{ NULL, 'r', NULL, 0,
	  N_("Output the real ID instead of the effective ID") },
	{ NULL, 'u', NULL, 0,
	  N_("Output only the effective user ID") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);

static bool opt_name;
static bool opt_real_id;
static enum opt_mode {
	ID_DEF,
	ID_GRP_ALL,
	ID_EGID,
	ID_EUID,
} opt_mode;
static char *opt_user;

static const struct argp argp = { options, parse_opt, args_doc, doc };

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'n':
		opt_name = true;
		break;
	case 'r':
		opt_real_id = true;
		break;
	case 'g':
		opt_mode = ID_EGID;
		break;
	case 'G':
		opt_mode = ID_GRP_ALL;
		break;
	case 'u':
		opt_mode = ID_EUID;
		break;

	case ARGP_KEY_ARG:
		if (state->arg_num > 0)		/* too many args */
			argp_usage (state);
		opt_user = arg;
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static int user_in_group (const struct group *gr)
{
	unsigned int i = 0;
	char *s;

	if (!opt_user)
		return 0;

	while ((s = gr->gr_mem[i++]) != NULL)
		if (!strcmp(s, opt_user))
			return 1;

	return 0;
}

struct grpent;
struct grpent {
	gid_t		gid;
	struct grpent	*next;
	char		name[0];
};

static void add_grpent(struct grpent **head, struct grpent **tail_io,
		      const struct group *gr)
{
	struct grpent *ent, *tail = *tail_io;

	ent = xmalloc(sizeof(struct grpent) + strlen(gr->gr_name) + 1);
	ent->gid = gr->gr_gid;
	ent->next = NULL;
	strcpy(ent->name, gr->gr_name);

	if (!*head)
		*head = ent;
	if (tail)
		tail->next = ent;
	*tail_io = ent;
}

static struct grpent *match_groups(gid_t gid, gid_t egid)
{
	struct group *gr;
	struct grpent *rc = NULL, *head = NULL, *tail = NULL;

	setgrent();

	while (1) {
		errno = 0;
		gr = getgrent();
		if (!gr) {
			if (errno) {
				perror("getgrent");
				goto out;
			}
			break;
		}

		if (gr->gr_gid == gid || gr->gr_gid == egid ||
		    user_in_group(gr)) {
			add_grpent(&head, &tail, gr);
		}
	}

	rc = head;

out:
	endgrent();
	return rc;
}

static void pr_uid(uid_t uid, bool space, bool newline)
{
	unsigned long long uid_ull;

	if (opt_name) {
		struct passwd *pw;

		errno = 0;
		pw = getpwuid(uid);
		if (!pw && errno)
			perror("warning(getpwuid)");
		if (pw) {
			printf("%s%s%s",
			       space ? " " : "",
			       pw->pw_name,
			       newline ? "\n" : "");
			return;
		}
	}

	uid_ull = uid;
	printf("%s%Lu%s",
	       space ? " " : "",
	       uid_ull,
	       newline ? "\n" : "");
}

static void pr_gid(gid_t gid, bool space, bool newline)
{
	unsigned long long gid_ull;

	if (opt_name) {
		struct group *gr;

		errno = 0;
		gr = getgrgid(gid);
		if (!gr && errno)
			perror("warning(getgrgid)");
		if (gr) {
			printf("%s%s%s",
			       space ? " " : "",
			       gr->gr_name,
			       newline ? "\n" : "");
			return;
		}
	}

	gid_ull = gid;
	printf("%s%Lu%s",
	       space ? " " : "",
	       gid_ull,
	       newline ? "\n" : "");
}

static int id_grp_all(struct grpent *grp, gid_t gid, gid_t egid)
{
	struct grpent *tmp;
	char *s, s1[64];

	tmp = grp;
	while (tmp) {
		if (opt_name)
			s = tmp->name;
		else {
			snprintf(s1, sizeof(s1), "%Lu",
				 (unsigned long long) tmp->gid);
			s = s1;
		}

		printf("%s%s%s",
		       (tmp == grp) ? "" : " ",
		       s,
		       (!tmp->next) ? "\n" : "");

		tmp = tmp->next;
	}

	return 0;
}

static int id_def(uid_t uid, uid_t euid, gid_t gid, gid_t egid,
		  struct grpent *grp)
{
	struct passwd *pw;
	struct group *gr;

	printf("uid=%Lu%s%s%s",
	       (unsigned long long) uid,
	       opt_user ? "(" : "",
	       opt_user ? opt_user : "",
	       opt_user ? ")" : "");

	if (uid != euid) {
		pw = getpwuid(euid);
		printf(" euid=%Lu%s%s%s",
		       (unsigned long long) euid,
		       pw ? "(" : "",
		       pw ? pw->pw_name : "",
		       pw ? ")" : "");
	}

	gr = getgrgid(gid);
	printf(" gid=%Lu%s%s%s",
	       (unsigned long long) gid,
	       gr ? "(" : "",
	       gr ? gr->gr_name : "",
	       gr ? ")" : "");

	if (gid != egid) {
		gr = getgrgid(egid);
		printf(" egid=%Lu%s%s%s",
		       (unsigned long long) egid,
		       gr ? "(" : "",
		       gr ? gr->gr_name : "",
		       gr ? ")" : "");
	}

	if (grp) {
		struct grpent *tmp = grp;
		printf(" groups=");

		while (tmp) {
			printf("%s%Lu(%s)",
			       (tmp == grp) ? "" : ",",
			       (unsigned long long) tmp->gid,
			       tmp->name);
			tmp = tmp->next;
		}
	}

	printf("\n");

	return 0;
}

static int do_id(void)
{
	uid_t uid, euid;
	gid_t gid, egid;
	struct passwd *pw;
	struct grpent *grp;

	if (opt_user) {
		errno = 0;
		pw = getpwnam(opt_user);
		if (!pw) {
			if (errno)
				perror(opt_user);
			else
				fprintf(stderr, "user '%s' not found\n",
					opt_user);
			return 1;
		}

		uid = euid = pw->pw_uid;
		gid = egid = pw->pw_gid;
	} else {
		uid = getuid();
		euid = geteuid();
		gid = getgid();
		egid = getegid();

		if (opt_real_id) {
			euid = uid;
			egid = gid;
		}

		if (opt_mode == ID_GRP_ALL || opt_mode == ID_DEF) {
			pw = getpwuid(uid);
			if (pw)
				opt_user = xstrdup(pw->pw_name);
		}
	}

	switch (opt_mode) {
	case ID_EUID:	pr_uid(euid, false, true); return 0;
	case ID_EGID:	pr_gid(egid, false, true); return 0;
	default:	break;
	}

	grp = match_groups(gid, egid);
	if (!grp)
		return 1;

	switch (opt_mode) {
	case ID_GRP_ALL:	return id_grp_all(grp, gid, egid);
	case ID_DEF:		return id_def(uid, euid, gid, egid, grp);
	default:		return 1;
	}
}

int main (int argc, char *argv[])
{
	error_t rc;

	pu_init();

	rc = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (rc) {
		fprintf(stderr, "argp_parse failed: %s\n", strerror(rc));
		return 1;
	}

	return do_id();
}

